### ISSUE TITLE
Installer error | "Redis server went away"

### ISSUE THREAD

**Issue by JHGitty** [mautic/mautic#979]
After input fields for PDO I get the following error:
```
 Uh oh! I think I broke it. If I do it again, please report me to the system administrator! 
```

I already tried latest dev and 1.2.0 release.

**Edit:**
**Error on 1.2.0 release and staging.**
```
mautic.CRITICAL: Uncaught PHP Exception RedisException: "Redis server went away"
```

**Comment by alanhartless** [comment]
Hi!

Can you please look in app/logs and possibly your server's logs to see if anything was recorded?

Thanks!

**Comment by JHGitty** [comment]
```
mautic.CRITICAL: Uncaught PHP Exception RedisException: "Redis server went away"
```
Redis? Is Mautic only working with Redis? Did I miss any official installation documentation?

**Comment by alanhartless** [comment]
No redis is not used by default and would only be used as a caching option for Doctrine but normally would require advanced/manual config file changes. Not sure why Doctrine is attempting to use it.  Is this with the code in staging?

**Comment by JHGitty** [comment]
The above error is with the code in staging, yes.
I'll try the same again with the 1.2.0 release and check which error will occur there.

**Comment by JHGitty** [comment]
*The same error occured on 1.2.0 release version*.
```
 mautic.CRITICAL: Uncaught PHP Exception RedisException: "Redis server went away" 
```

**Comment by mbabker** [comment]
My local environment has Redis installed, so I went ahead and ran through the Mautic installer a couple of times (in both production and dev environments) and the only reference I can find to Redis in that workflow is in Symfony's compiled container that gets written to the cache directory.  This is only the default configuration values for the services, the same way as all other Symfony bundles work.  If the application were trying to use Redis, the internal methods to build the services would be referencing that configuration and from what I can see that doesn't happen.  For example...

```php
    protected function getDoctrineCache_Providers_Doctrine_Orm_DefaultMetadataCacheService()
    {
        $this->services['doctrine_cache.providers.doctrine.orm.default_metadata_cache'] = $instance = new \Doctrine\Common\Cache\ArrayCache();
        $instance->setNamespace('sf2orm_default_857ccabf4a15a6bb46ba54f94e8fd86795503ffc6d1102b492a4a4c4bdbdef60');
        return $instance;
    }
```

As for your specific case, do you have access to PHP's error logs or the Apache logs?  More data about what is throwing the error may be available in those logs which would help to figure out what's trying to call Redis based code.

**Comment by JHGitty** [comment]
@mbabker On our system Redis is *not installed*. Redis should not be used by default.
I'll look into it, currently I cannot find any more error information on the other log files.

**Comment by JHGitty** [comment]
```
mod_fcgid: stderr: PHP Fatal error:  require_once(): Failed opening required '/var/www/vhosts/ads.emedia-marketing.de/technik-oase.de/marketing/app/bootstrap.php.cache'
```
I've found this error on PHP error_logs. How can I fix it?
The above error is from 1.2.0 release.

**Comment by alanhartless** [comment]
Installing from Git requires extra steps namely, `composer install`.  See https://github.com/mautic/mautic#getting-started.

**Comment by JHGitty** [comment]
@alanhartless The last error above is from 1.2.0 release from website, not from git.
On staging (git version) I used ``composer install``.

**Comment by JHGitty** [comment]
Ok, let's only talk about 1.2.0 release version, which I've downloaded from the website.

1. I should not use ``composer install``. (Btw there is no composer.json on release version)
2. I get above ``Redis gone away`` error on app/logs prod log
3. I get above ``Failed opening required ... bootstrap.php.cache`` on php error logs
4. After installer step for database (I use PDO) I get ``500 Internal Server Error``.

Any idea why?

**Comment by alanhartless** [comment]
With the 1.2.0 release version from mautic.org/download, no, you should not have to run composer install.  But it should also have the bootstrap.php.cache file with it so if you're getting a fatal error that it is missing, then maybe look to ensure it was uploaded and that read permissions are correct.

By "get above," do you mean you've fixed those issues and are no longer getting the errors?

The 500 internal server error could be anything.  Check your server's PHP error log and Mautic's logs in app/logs to see if anything is recorded that correlates with the 500.

**Comment by JHGitty** [comment]
@alanhartless I think I've done a mistake about the 3rd error above. The "bootstrap.php.cache" error was from staging before I've done ``composer install`` I think.

Ok, so only let's talk about 1.2.0 release. There is the ``Redis gone away`` error - **but why?**

With "I get above error" I mean that it is the same error as I wrote in the last messages. Sorry for my bad english.

There are no more errors on PHP error log or Mautic's logs. Only the things I've already wrote.

I'll check the permissions of all files and try it again with 777 on all files to be sure that the permissions are not the reason.

**Comment by JHGitty** [comment]
I've set chmod 777 to *all files of Mautic*. Not working..

There are no errors on the PHP error log. The only thing I've found is on Mautic's logs:
```
 mautic.CRITICAL: Uncaught PHP Exception RedisException: "Redis server went away" 
```

**Comment by alanhartless** [comment]
If you think you've got mixed files, I would delete anything installed from staging and just upload the files in the release package to see if you get the same results.

The redis gone away error is definitely a new one for us and a bit of a mystery at the moment. We'll have to dig into that.

**Comment by JHGitty** [comment]
@alanhartless There are **no mixed files**. I've always deleted the folders to be sure.
If you like you can contact me privately to check this. I am not able to bypass the ``Redis server went away`` error. I don't know why it happens.

**Comment by mbabker** [comment]
This might help get some debug info, no promises though.

In `app/bundles/CoreBundle/Controller/ExceptionController.php`, add this code snippet to the beginning of the `showAction()` method:

```php
$this->get('monolog.logger.mautic')->addError($exception->getMessage(), array('trace' => $exception->getTrace()));
```

With any luck, that will cause an additional entry to be written to the log file with a stack trace anytime an Exception is thrown and goes uncaught (which is the case with the RedisException here).

**Comment by JHGitty** [comment]
@mbabker

It returns a stacktrace in the log file:
```json
mautic.ERROR: Redis server went away {
    "trace": [{
        "namespace": "",
        "short_class": "",
        "class": "",
        "type": "",
        "function": "",
        "file": "(absolute path)vendor/doctrine/cache/lib/Doctrine/Common/Cache/RedisCache.php",
        "line": 47,
        "args": []
    }, {
        "namespace": "",
        "short_class": "Redis",
        "class": "Redis",
        "type": "->",
        "function": "setOption",
        "file": "(absolute path)vendor/doctrine/cache/lib/Doctrine/Common/Cache/RedisCache.php",
        "line": 47,
        "args": [
            ["string", "1"],
            ["string", "1"]
        ]
    }, {
        "namespace": "Doctrine\\Common\\Cache",
        "short_class": "RedisCache",
        "class": "Doctrine\\Common\\Cache\\RedisCache",
        "type": "->",
        "function": "setRedis",
        "file": "(absolute path)vendor/doctrine/orm/lib/Doctrine/ORM/Tools/Setup.php",
        "line": 139,
        "args": [
            ["object", "Redis"]
        ]
    }, {
        "namespace": "Doctrine\\ORM\\Tools",
        "short_class": "Setup",
        "class": "Doctrine\\ORM\\Tools\\Setup",
        "type": "::",
        "function": "createConfiguration",
        "file": "(absolute path)app/bundles/InstallBundle/Controller/InstallController.php",
        "line": 904,
        "args": []
    }, {
        "namespace": "Mautic\\InstallBundle\\Controller",
        "short_class": "InstallController",
        "class": "Mautic\\InstallBundle\\Controller\\InstallController",
        "type": "->",
        "function": "getEntityManager",
        "file": "(absolute path)app/bundles/InstallBundle/Controller/InstallController.php",
        "line": 521,
        "args": [
            ["array", {
                "driver": ["string", "pdo_mysql"],
                "host": ["string", "localhost"],
                "table_prefix": ["null", null],
                "port": ["string", "3306"],
                "user": ["string", "Mautic_Marketing"],
                "password": ["string", "nMqt749#"],
                "path": ["null", null],
                "backup_tables": ["boolean", false],
                "backup_prefix": ["string", "bak_"],
                "dbname": ["string", "Mautic_Marketing"],
                "charset": ["string", "UTF8"]
            }]
        ]
    }, {
        "namespace": "Mautic\\InstallBundle\\Controller",
        "short_class": "InstallController",
        "class": "Mautic\\InstallBundle\\Controller\\InstallController",
        "type": "->",
        "function": "performDatabaseInstallation",
        "file": "(absolute path)app/bundles/InstallBundle/Controller/InstallController.php",
        "line": 174,
        "args": [
            ["array", {
                "driver": ["string", "pdo_mysql"],
                "host": ["string", "localhost"],
                "table_prefix": ["null", null],
                "port": ["string", "3306"],
                "user": ["string", "Mautic_Marketing"],
                "password": ["string", "nMqt749#"],
                "path": ["null", null],
                "backup_tables": ["boolean", false],
                "backup_prefix": ["string", "bak_"],
                "dbname": ["string", "Mautic_Marketing"]
            }]
        ]
    }, {
        "namespace": "Mautic\\InstallBundle\\Controller",
        "short_class": "InstallController",
        "class": "Mautic\\InstallBundle\\Controller\\InstallController",
        "type": "->",
        "function": "stepAction",
        "file": null,
        "line": null,
        "args": [
            ["string", "1"]
        ]
    }, {
        "namespace": "",
        "short_class": "",
        "class": "",
        "type": "",
        "function": "call_user_func_array",
        "file": "(absolute path)app/bootstrap.php.cache",
        "line": 3051,
        "args": [
            ["array", [
                ["object", "Mautic\\InstallBundle\\Controller\\InstallController"],
                ["string", "stepAction"]
            ]],
            ["array", [
                ["string", "1"]
            ]]
        ]
    }, {
        "namespace": "Symfony\\Component\\HttpKernel",
        "short_class": "HttpKernel",
        "class": "Symfony\\Component\\HttpKernel\\HttpKernel",
        "type": "->",
        "function": "handleRaw",
        "file": "(absolute path)app/bootstrap.php.cache",
        "line": 3013,
        "args": [
            ["object", "Symfony\\Component\\HttpFoundation\\Request"],
            ["string", "1"]
        ]
    }, {
        "namespace": "Symfony\\Component\\HttpKernel",
        "short_class": "HttpKernel",
        "class": "Symfony\\Component\\HttpKernel\\HttpKernel",
        "type": "->",
        "function": "handle",
        "file": "(absolute path)app/bootstrap.php.cache",
        "line": 3162,
        "args": [
            ["object", "Symfony\\Component\\HttpFoundation\\Request"],
            ["string", "1"],
            ["boolean", true]
        ]
    }, {
        "namespace": "Symfony\\Component\\HttpKernel\\DependencyInjection",
        "short_class": "ContainerAwareHttpKernel",
        "class": "Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel",
        "type": "->",
        "function": "handle",
        "file": "(absolute path)app/bootstrap.php.cache",
        "line": 2403,
        "args": [
            ["object", "Symfony\\Component\\HttpFoundation\\Request"],
            ["string", "1"],
            ["boolean", true]
        ]
    }, {
        "namespace": "Symfony\\Component\\HttpKernel",
        "short_class": "Kernel",
        "class": "Symfony\\Component\\HttpKernel\\Kernel",
        "type": "->",
        "function": "handle",
        "file": "(absolute path)app/AppKernel.php",
        "line": 110,
        "args": [
            ["object", "Symfony\\Component\\HttpFoundation\\Request"],
            ["string", "1"],
            ["boolean", true]
        ]
    }, {
        "namespace": "",
        "short_class": "AppKernel",
        "class": "AppKernel",
        "type": "->",
        "function": "handle",
        "file": "(absolute path)index.php",
        "line": 56,
        "args": [
            ["object", "Symfony\\Component\\HttpFoundation\\Request"]
        ]
    }]
}[]
```

**Comment by mbabker** [comment]
So based on that, what I think is happening is that your environment has PHP's Redis extension installed (this can be checked with a `phpinfo()` dump) but you don't have a running Redis instance which causes the Exception to be thrown.  By default, the method that's trying to build the Redis based object checks if a Cache object was injected into it and if not tries to build one by checking what PHP extensions are enabled (in order; APC, XCache, Memcache, and Redis) and if one of those is enabled it builds an object for that extension, and falls back to a default object otherwise.

In PR #983 I've adjusted the code to inject that default object so it bypasses the check.  That should get around the issue for you.

**Comment by JHGitty** [comment]
@mbabker Confirmed. Plesk 12.5 automatically installs and enables the `redis` extension for all PHP instances **by default**. We are not running any Redis instance, you are right!

It is important that there is a fallback or check to bypass this issue by default because a lot of server providers / hosters / users / .. could get this issue on their setup.

We will try to bypass it by disabling `redis` PHP extension on our system.

**Comment by alanhartless** [comment]
@JHGitty You could try patching the code with Michael's PR if you wanted.  That should prevent the error as well and provides the fallback you're referring to.  It'll be in the next release.

Thanks for helping us find that! Sorry you had to go through the frustration in doing so though.

**Comment by JHGitty** [comment]
We not tried the PR yet, but removing PHP extension 'redis' worked. :-)

**Comment by alanhartless** [comment]
Fixed in #983.
